hmmlearn/hmm.py

```diff
@@ -520,134 +525,294 @@ class GMMHMM(_BaseHMM):
+                             (np.ones((self.n_components, 1)) * self.n_mix))
+
+        if 'm' in self.init_params or not hasattr(self, "means_"):
+            self.means_ = np.zeros((self.n_components, self.n_mix,
+                                    self.n_features))
+            for i, kmeans in enumerate(kmeanses):
+                self.means_[i] = kmeans.cluster_centers_
+
+        if 'c' in self.init_params or not hasattr(self, "covars_"):
+            cv = np.cov(X.T) + self.min_covar * np.eye(self.n_features)
+            if not cv.shape:
+                cv.shape = (1, 1)
+
+            if self.covariance_type == 'tied':
+                self.covars_ = np.zeros((self.n_components,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'full':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features, self.n_features))
+                self.covars_[:] = cv
+            elif self.covariance_type == 'diag':
+                self.covars_ = np.zeros((self.n_components, self.n_mix,
+                                         self.n_features))
+                self.covars_[:] = np.diag(cv)
+            elif self.covariance_type == 'spherical':
+                self.covars_ = np.zeros((self.n_components, self.n_mix))
+                self.covars_[:] = cv.mean()
+
+    def _check(self):
+        super(GMMHMM, self)._check()
+
+        if not hasattr(self, "n_features"):
+            self.n_features = self.means_.shape[2]
+
+        # Checking covariance type
+        if self.covariance_type not in COVARIANCE_TYPES:
+            raise ValueError("covariance_type must be one of {0}"
+                             .format(COVARIANCE_TYPES))
+
+        self.weights_ = np.array(self.weights_)
+        # Checking mixture weights' shape
+        if self.weights_.shape != (self.n_components, self.n_mix):
+            raise ValueError("mixture weights must have shape "
+                             "(n_components, n_mix), "
+                             "actual shape: {}".format(self.weights_.shape))
+
+        # Checking mixture weights' mathematical correctness
+        if not np.allclose(np.sum(self.weights_, axis=1),
+                           np.ones(self.n_components)):
+            raise ValueError("mixture weights must sum up to 1")
+
+        # Checking means' shape
+        self.means_ = np.array(self.means_)
+        if self.means_.shape != (self.n_components, self.n_mix,
+                                 self.n_features):
+            raise ValueError("mixture means must have shape "
+                             "(n_components, n_mix, n_features), "
+                             "actual shape: {}".format(self.means_.shape))
+
+        # Checking covariances' shape
+        self.covars_ = np.array(self.covars_)
```

Comment:
I would agree if any data was actually flowing in this code (and I did exactly this in other parts of code, where data is actually passed around), but here this is basically a script. No data, we just need to execute these lines of code. If I were to extract the code in separate functions, these functions won't take or return any values. And if, say, you use some IDE, which can fold custom blocks of code, commented block of code would be literally the same as separate function. Even if you don't have such a feature, I don't think scrolling is that big of a task.

> Consider a hundred lines long function. To get an idea of how it works, you need to skim through all this lines.

You really don't, if there are proper comments in place. Sure, comments get outdated, but so do function names. And in the case as simple as this, I don't think keeping comments up to date is any harder than function names. In fact, it can be way simpler, if I'm not using IDE with "Rename function" feature.

And the advantage of not using function in our case is that if I want to really understand what the code does, I don't need to hop around the source file, I can just see it all in one place, where it is actually executed. Also, Python does not have private methods, so the namespace is not getting polluted.